### PR TITLE
[language][move-cli] remove diem dependencies

### PR DESCRIPTION
**Pull request by vgao1996**
## Summary
This makes the move-cli no longer depend on Diem crates. 
- The `move-cli` crate has been made a library. Callers must provide native functions and errmaps.
  - The `move` binary now calls into the library, with the Move stdlib natives and error map.
  - A new binary named `move-df` is introduced, which lives in a separate crate called `move-cli-with-df`. This binary has access to Diem natives and error map.
- The `INITIAL_GAS_SCHEDULE` is moved from `vm-genesis` into `move-vm-types`.

See individual commits for more details.

## Dependencies
#8584

### CODE REVIEW

File: language/move-vm/types/src/gas_schedule.rs [hunk 1]
```diff
@@ -277,6 +277,145 @@ pub fn zero_cost_schedule() -> CostTable {
     new_from_instructions(instrs, native_table)
 }
 
+pub static INITIAL_GAS_SCHEDULE: Lazy<CostTable> = Lazy::new(|| {
```
Comment (tnowacki) [inline language/move-vm/types/src/gas_schedule.rs:280]: Was the other one removed from vm_genesis? Maybe we can comment what this one is for?

**Review by tnowacki — COMMENTED** [review]
Before I keep reviewing, it looks like a bunch of commits got included with this and I might be reviewing old stuff or already landed items. Can you point me at the things that are just for this PR?

**Comment by vgao1996** [comment]
@tnowacki the first commit in the stack (the resource viewer one) belongs to #8584, which is landing right now. All other commits are new changes I'm introducing. Each of those commits (except the last one) removes a single diem dependency so it might be more convenient if you read them in isolation.

**Comment by tnowacki** [comment]
> All other commits are new changes I'm introducing. Each of those commits (except the last one) removes a single diem dependency so it might be more convenient if you read them in isolation.

I'll try to take a look through each one. But some of these don't seem really related to the cli :p

File: language/tools/move-cli/src/lib.rs [hunk 2]
```diff
@@ -22,3 +22,344 @@ pub use move_lang::command_line::DEFAULT_OUTPUT_DIR as DEFAULT_BUILD_DIR;
 
 /// Extension for resource and event files, which are in BCS format
 const BCS_EXTENSION: &str = "bcs";
+
```
Comment (tnowacki) [inline language/tools/move-cli/src/lib.rs:25]: can you leave this in main.rs. It seems unrelated to cleaning up deps

File: language/tools/move-cli-with-df/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+[package]
```
Comment (vgao1996) [inline language/tools/move-cli-with-df/Cargo.toml:1]: BTW, the reason why I think it's necessary to introduce a separate crate is native functions. In theory, error code maps or Move source code can be specified via command line options, or a config file if we find that tedious, but not native functions.

File: language/move-stdlib/src/lib.rs [hunk 4]
```diff
@@ -146,3 +151,25 @@ pub fn build_nursery_doc(output_path: &str) {
         false,
     )
 }
+
+pub fn build_error_code_map(output_path: &str) {
```
Comment (tnowacki) [inline language/move-stdlib/src/lib.rs:155]: @tzakian can you double check this

File: language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs [hunk 5]
```diff
@@ -112,10 +113,16 @@ impl OnDiskStateView {
     }
 
     // Events are stored under address/handle creation number
-    fn get_event_path(&self, key: &EventKey) -> PathBuf {
-        let mut path = self.get_addr_path(&key.get_creator_address());
+    fn get_event_path(&self, key: &[u8]) -> PathBuf {
+        // TODO: this is a hacky way to get the account address and creation number from the event key.
```
Comment (tnowacki) [inline language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs:117]: Is this hacky? The Event module is in stdlib not the diem framework

File: language/move-vm/types/src/gas_schedule.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-vm/types/src/gas_schedule.rs:280]: cc @tzakian

**Review by tnowacki — CHANGES_REQUESTED** [review]
requesting changes for the main.rs ~> lib.rs thing. Makes it really hard to see what is happening

File: language/tools/move-cli/src/lib.rs (hunk 2 above)
Comment (vgao1996) [reply language/tools/move-cli/src/lib.rs:25]: Unfortunately it does. Because of native functions I had to introduce a new crate `move-cli-with-df` and a new binary `move-df`. They need to share the same piece CLI code, which I moved from main.rs into lib.rs.

File: language/move-stdlib/src/lib.rs (hunk 4 above)
Comment (tzakian) [reply language/move-stdlib/src/lib.rs:155]: LGTM!

File: language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs (hunk 5 above)
Comment (vgao1996) [reply language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs:117]: It's hacky in the sense that our current definition of events is way too Diem-specific, and what I'm doing here is effectively duplicating the logic in `diem_types::event::EventKey`. We will deal with this once we make (Move) events more generic.

Comment (tnowacki) [reply language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs:117]: I suppose a bit with the events dir + native function. But the layout of the key is still defined in the the Events module which seems okay to me

File: language/move-stdlib/src/lib.rs (hunk 4 above)
Comment (vgao1996) [reply language/move-stdlib/src/lib.rs:155]: I wonder if we should also include the nursery modules

File: language/move-vm/types/src/gas_schedule.rs (hunk 1 above)
Comment (tzakian) [reply language/move-vm/types/src/gas_schedule.rs:280]: I don't think this should exist here. This gas schedule is Diem-specific, so doesn't really make sense to be in the move-vm IMO. 

If you need a gas schedule I would recommend using either the zero-cost cost table, or a unit cost table (e.g., https://github.com/diem/diem/blob/main/language/tools/move-unit-test/src/test_runner.rs#L58).

File: language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs (hunk 5 above)
Comment (vgao1996) [reply language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs:117]: > But the layout of the key is still defined in the the Events module which seems okay to me

Yeah, this is not a problem under the current definition given by the Events module. Just saying that this definition itself is Diem-specific and we'll likely need to change it to something more general, in order to support alternative clients like Trove.

File: language/move-vm/types/src/gas_schedule.rs (hunk 1 above)
Comment (vgao1996) [reply language/move-vm/types/src/gas_schedule.rs:280]: > I don't think this should exist here. This gas schedule is Diem-specific, so doesn't really make sense to be in the move-vm IMO.

That's reasonable. I did this because I didn't know better alternatives. BTW `pub enum NativeCostIndex` suffers from the same problem -- some of its entries are diem-framework natives. I think we should move it out of `move-vm-types`.

> If you need a gas schedule I would recommend using either the zero-cost cost table, or a unit cost table

The `unit-cost-table` sounds appealing to me. I guess it's good enough for the cli for testing purposes?

Comment (vgao1996) [reply language/move-vm/types/src/gas_schedule.rs:280]: The broader issue here is that native functions uses `NativeCostIndex` to determine their costs. This is problematic since these indices could be different for different deployments and different combinations of packages with native functions, at least in theory. In other words, I think we'll need to figure out a way to make the native cost table composable, or behave more like unordered sets in the future.

File: language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs [hunk 6]
```diff
@@ -4,7 +4,6 @@
 use crate::{BCS_EXTENSION, DEFAULT_BUILD_DIR, DEFAULT_STORAGE_DIR};
 use disassembler::disassembler::Disassembler;
 // TODO: do we want to make these Move core types or allow this to be customizable?
```
Comment (sblackshear) [inline language/tools/move-cli/src/sandbox/utils/on_disk_state_view.rs:6]: Can delete this TODO now

**Comment by sblackshear** [comment]
> A new binary named move-df is introduced, which lives in a separate crate called move-cli-with-df. This binary has access to Diem natives and error map.

Ideally, we wouldn't need to set up a new crate/binary for each "instantiation" of the CLI with natives, framework, and adapter... I wonder if there are other ways that we could package these artifacts together.

**Comment by vgao1996** [comment]
> Ideally, we wouldn't need to set up a new crate/binary for each "instantiation" of the CLI with natives, framework, and adapter... I wonder if there are other ways that we could package these artifacts together.

Great question! Here's what I think
- Framework?
  - Definitely possible. Requiring paths to be specified via command line options would work, despite being a less than ideal user experience. To make it better, we could consider 
    - Adding a config file to free users from having to type the path every time, undoing the inconvenience
    - Better defining our package format and making the cli deal with generic packages rather than specific ones like the move-stdlib or diem-framework
- Natives?
  - This will be tough since our natives are implemented in Rust, a compiled language. If we were to make this configurable, I'd assume some sort of dynamic loading mechanism (.so, .dll etc.) would be required. 
- Adapter?
  - Still technically doable, but probably as difficult or even harder than natives since an alternative adapter may require a change in the execution flow, and it may also want to add commands that are specific to it.

File: x.toml [hunk 7]
```diff
@@ -256,12 +257,6 @@ existing_deps = [
     # Tier 0 - essential components that must be included in the first batch of crates getting
     #          moved to the new repo
     ["move-lang", "diem-framework"],                          # sanity check to ensure that the diem-framework compiles
-    ["compiler", "diem-framework-releases"],
```
Comment (vgao1996) [inline x.toml:259]: I forgot to delete this in #8576. Doing it here just to save myself from some more merge conflicts

**Comment by sblackshear** [comment]
> > Ideally, we wouldn't need to set up a new crate/binary for each "instantiation" of the CLI with natives, framework, and adapter... I wonder if there are other ways that we could package these artifacts together.
> 
> Great question! Here's what I think
> 
> * Framework?
>   
>   * Definitely possible. Requiring paths to be specified via command line options would work, despite being a less than ideal user experience. To make it better, we could consider
>     
>     * Adding a config file to free users from having to type the path every time
>     * Better defining our package format and making the cli use generic packages rather than hardcoded ones like the move-stdlib or diem-framework
> * Natives?
>   
>   * This will be tough since our natives are implemented in Rust, a compiled language. If we were to make this configurable, I'd assume some sort of dynamic loading mechanism (.so, .dll etc.) would be required. In addition, the cli will likely need to be OS-aware so it can handle dynamic loading correctly.
> * Adapter?
>   
>   * This depends on your definition of pluggable adapter.
>     
>     * If all you need is calling some prologue or epilogue functions, then we may be able to get away with a description file.
>     * If the adapter requires a more significant change in the execution flow, or wants to add a new command that's specific to it, then I feel like a separate binary might be a better approach.

Good thoughts!

Agreed that natives and the adapter are harder problems due to Rust's lack of dynamism. Dynamic loading at a lower level would be interesting, but I think your solution is a good one for now.

File: Cargo.toml [hunk 8]
```diff
@@ -113,6 +113,7 @@ members = [
     "language/tools/move-bytecode-utils",
     "language/tools/move-bytecode-viewer",
     "language/tools/move-cli",
+    "language/tools/move-cli-with-df",
```
Comment (sblackshear) [inline Cargo.toml:116]: Could we move this to `diem-tools` and call it something snapper like `df-cli`?

**Comment by vgao1996** [comment]
@sblackshear I renamed `move-cli-with-df` to `df-cli` and moved it into diem-tools. Let me know if it looks good to you now!

File: language/tools/move-cli/src/main.rs [hunk 9]
```diff
@@ -2,339 +2,13 @@
...
-            }
-            Ok(())
-        }
-        SandboxCommand::Doctor {} => {
-            let state = mode.prepare_state(&move_args.build_dir, &move_args.storage_dir)?;
-            sandbox::commands::doctor(&state)
-        }
-    }
+    let error_descriptions: ErrorMapping =
+        bcs::from_bytes(move_stdlib::error_descriptions()).unwrap();
```
Comment (sblackshear) [inline language/tools/move-cli/src/main.rs:9]: Since `main` returns `Result`, should we replace this `unwrap`+ the one below with `?`'s?

File: language/tools/move-cli/src/sandbox/commands/publish.rs [hunk 10]
```diff
@@ -1,16 +1,20 @@
...
+    },
+    NativeFunctionRecord,
 };
 use move_lang::{self, compiled_unit::CompiledUnit, Compiler, Flags};
 use move_vm_runtime::move_vm::MoveVM;
 
 use anyhow::Result;
 
 pub fn publish(
+    natives: Vec<NativeFunctionRecord>,
```
Comment (sblackshear) [inline language/tools/move-cli/src/sandbox/commands/publish.rs:17]: Nit: consider `IntoIterator` or a slice (since `MoveVM` doesn't need ownership?)

File: language/tools/move-cli/src/sandbox/commands/run.rs [hunk 11]
```diff
@@ -18,6 +22,8 @@ use move_vm_runtime::move_vm::MoveVM;
 use anyhow::{anyhow, bail, Result};
 use std::{fs, path::Path};
 pub fn run(
+    natives: Vec<NativeFunctionRecord>,
```
Comment (sblackshear) [inline language/tools/move-cli/src/sandbox/commands/run.rs:25]: Similar here.

File: language/tools/move-cli/src/sandbox/utils/mod.rs [hunk 12]
```diff
@@ -367,11 +368,6 @@ pub(crate) fn explain_execution_error(
     use StatusCode::*;
     match error.into_vm_status() {
         VMStatus::MoveAbort(AbortLocation::Module(id), abort_code) => {
-            // try to use move-explain to explain the abort
```
Comment (sblackshear) [inline language/tools/move-cli/src/sandbox/utils/mod.rs:370]: Keep this comment?

**Review by sblackshear — DISMISSED** [review]

File: language/tools/move-cli/src/main.rs [hunk 13]
```diff
@@ -2,339 +2,13 @@
...
+    move_cli::move_cli(
+        move_stdlib::natives::all_natives(AccountAddress::from_hex_literal("0x1").unwrap()),
```
[the remark this answers is not in the document]
Comment (vgao1996) [reply language/tools/move-cli/src/main.rs:11]: I don't think so. With stdlib now using named addresses, the actual numeric binding is technically determined by each deployment. Let's wait and see how the package system rolls out and how that would affect our handling of addresses in various deployments.

File: language/tools/move-cli/src/sandbox/commands/publish.rs (hunk 10 above)
Comment (vgao1996) [reply language/tools/move-cli/src/sandbox/commands/publish.rs:17]: Well the VM does need ownership here. This record contains fully qualified function names that will be used as keys in collections. With that being said, I'll switch it to `impl IntoIterator<NativeFunctionRecord>`, just to make it more general.

**Review by sblackshear — APPROVED** [review]
